### DISCUSSION TITLE
Revise statecharts/re-frame integration

### DISCUSSION THREAD

**Discussion by ingesolvoll** [ingesolvoll/glimt#4] [Ideas]
This is a continuation thread for the discussion going on in #3 

@mainej and I are discussing alternatives to the current integration, with the following potential problems to solve:
- Problem of scaling caused by too many FSM declarations, events or interceptors
- Potential memory leak caused by completed FSMs leaving their state in client db, and their now unused events/interceptors in the registry.
- State and machine are still a bit mixed together in the current clj-statecharts integration
- https://github.com/ingesolvoll/kee-frame has a custom clj-statecharts re-frame integration going, and it also integrates with this library. I would ideally want those to play nice together, and use the same integration in a composable way.

Lots of more detail to start with in the PR.

**Comment by mainej** [comment]
@ingesolvoll—thanks for putting so much thought into this question. 

I started to experiment with some code before reading your responses.

Here where I got on [the `glimt` side](https://github.com/mainej/glimt/compare/main...re-frame-multi). And here are the supporting [changes to `clj-statecharts`](https://github.com/mainej/clj-statecharts/compare/master...re-frame-multi). Here's [a gist](https://gist.github.com/mainej/c9621864d3cb7423a1efc8942d65a606) of using `glimt.multi-state`, the new `glimt` namespace.

I like where I got with `clj-statecharts`. The code persists both the machines and states in `re-frame`'s app-db separately. It doesn't register event handlers for an individual state machine. Instead, it has a single `::initialize` handler and a single `::transition` handler. Those handlers expect to be given enough information in their event vectors to look up the right machine and state.

I think this approach is similar to what `clj-statecharts` calls a "service". That is, it uses the app-db as a mutable stateful database of immutable machines and states.

The `glimt` changes are, I hope, unsurprising. The approach is different enough that I created a new namespace. As far as implementation goes, it's mostly a matter of dispatching `clj-statecharts` event vectors which contain both a machine id and a state id.

I split up the code for registering a machine and for starting a request before reading your comment about whether the `glimt` API should expose 1 or 2 events. I agree, a narrow API can be better for developer ergonomics. But, I've also seen many other re-frame libraries that require some setup at app-load, or when some segment of the app is first visited. In any case, I think it would be possible to bring the `glimt` API back down to 1 event. If we get further with this plan, I can look into that. On the other hand, maybe seeing it convinces you it's OK as 2 events.

Regarding where to publish the changes... Unfortunately, to make the `clj-statecharts` code work, I had to make one change deep down in its implementation. (That change is on [this line](https://github.com/mainej/clj-statecharts/compare/master...re-frame-multi#diff-72b1edc6885b8fa120f60bd17e6eafa960032a266179b2a199cd89f6893175d1R232). It allows a scheduler to be on a state, so that retries have enough information to resume with a particular state, not just a machine.) I think that change precludes publishing a separate library. At least that one piece would have to be merged back into the main `clj-statecharts` repo. (Or I'd have to maintain a fork, which I'm not excited about.) As you said, we don't know whether the `clj-statecharts` maintainer will be amenable to PRs, though I'm willing to give it a shot. If you can see another way to do it without modifying the `clj-statecharts` implementation, let me know.

I don't think the kee-frame solution would work for me as is. I like that it stores the state in the db, but it creates a one-to-one correspondence between a state and a machine. It also registers a new event handler for every machine/state combo, closing over the machine as it does so. That creates memory leaks in the global event handler registry when you have thousands of requests.

Some other notes on the experiment so far:

* It was pretty easy to implement `::discard` and `::restart` in the new `glimt.multi-state` namespace. I can imagine moving `::discard-state` to `clj-statecharts`, but that's a minor detail.
* In `clj-statecharts` I haven't implemented `::unregister` to discard a machine yet although it looks like it will be easy. The code is already set up to drop transitions if the machine can't be found. Fortunately, there's no longer any need to un-register event handlers or worry about what happens if non-existent handlers get triggered by retries.
* In `clj-statecharts` I stripped out the "epoch" idea though I plan to add it back in. It doesn't seem to be important for `glimt`, and I think I see a much cleaner way to implement it (by storing an epoch on the machine, and on each state.)
* I'm not sure I follow your idea about having a single machine for all of `glimt`. As I understand it, your thinking is that even things like the retry policies would be stored in the state? Compared to what I have now, it would be slightly easier to refer to 1 single machine, instead of looking one up from the app-db. But I think that single machine would be more complicated, because it would have to deduce much of its operation from the state. Tradeoffs... I'll have to think about it. At the moment, I don't mind managing a small collection of state machines. (Actually, I can see some benefit to having separate machines for different sections of an app. When combined with the above idea, when you leave one part of the app, you could `::unregister` that section's machine, and thereby drop all of its in-flight requests.)
* Something weird was happening with dependencies when I switched to working on these two repos locally. `malli` started throwing [an error](https://github.com/metosin/malli/commit/98167660c841fdf9618974a83e7e7d2fa3acd5b2). To fix that error I updated the dep on `malli` in `clj-statecharts`, but that caused some bugs in `glimt`. This [recent Issue](https://github.com/metosin/malli/issues/351) and the [associated PR](https://github.com/metosin/malli/pull/397/files) mean that malli no longer really works when you have map keys that are both `:optional` and have a `:default`. The keys now get dropped instead of defaulted. I hacked around it in `glimt`, but I'd want to come back to that area of code. Whether my code ever gets merged into `glimt`, you should be aware that `glimt.core` will run into these errors whenever `clj-statecharts` updates `malli`.

**Comment by ingesolvoll** [comment]
@mainej This is great, thank you! I don't have time to process all the details right now, but will do later today.

Why don't you  create a PR in glimt for your changes, and we can discuss the details there.

I'll see if I can bring Lucy Wang and possibly other people in here to discuss the rest.

**Comment by ingesolvoll** [comment]
https://clojurians.slack.com/archives/C01C7RJA81M/p1635148972001100

**Comment by mainej** [comment]
@ingesolvoll I can create a PR for `glimt`, although it wouldn't be mergable until `clj-statecharts` has been modified and released. 

The `clj-statecharts` code is nearly ready for an initial review. It'll need tests and documentation, among other things. The part that I'm worried about is the bump to `malli`, especially since I know it caused errors in `glimt`. I suppose `clj-statecharts` may want to do that upgrade eventually, but I don't think it belongs in the scope right now. So, I'm going to work on downgrading `malli` in `clj-statecharts`, though that'll need more debugging in `glimt`.

A small progress update... in `clj-statecharts` I implemented `::unregister` (which discards a machine) and `::discard-state` (which discards a state). I also borrowed the idea of "epochs", though I think I found a simpler (to me at least!) implementation. You can see those changes in the link I provided earlier: [`clj-statecharts`](https://github.com/mainej/clj-statecharts/compare/master...re-frame-multi). Those `clj-statecharts` additions only lead to a trivial change to `glimt`.

**Reply by mainej** [reply]
> work on downgrading `malli`

This was tricky, but I think I have it working. The released version of `clj-statecharts` on clojars uses malli 0.3.0. If `clj-statecharts` were released today, it [looks like](https://github.com/lucywang000/clj-statecharts/blob/df37e27bde149b0a64b65a5f4390ce21a0431549/project.clj#L20) it would use 0.6.1, though that version is buggy and the latest version is 0.6.2. I had a hard time pinning my test code to a particular version of malli, but once I was able to, I confirmed that the schemas in the new code work OK with either 0.3.0 or 0.6.2. 

The schema in `glimt.core` currently only works with `malli` 0.3.0. I know how to fix it, but want to keep that change out of the current discussion. Regardless of what happens with this new code, I can help fix problems when that upgrade eventually happens. The basic idea is that you need two schemas: one for validation and one for decoding. You can see an example of that in the new code.

Now that the dependency issues are worked out, I'll work on PRs for`clj-statecharts` and `glimt`. We can continue the discussion there.

**Comment by mainej** [comment]
clj-statecharts PR: https://github.com/lucywang000/clj-statecharts/pull/7

@ingesolvoll I'll work on starting a `glimt` PR, but may not have it written until tomorrow (I'm on Pacific time in the US).

**Comment by mainej** [comment]
glimt PR: https://github.com/ingesolvoll/glimt/pull/5
